clash-ghc/src-ghc/Clash/GHC/LoadInterfaceFiles.hs

```diff
@@ -103,49 +113,127 @@ import qualified Clash.Util.Interpolate              as I
 
 -- | Data structure tracking loaded binders (and their related data)
 data LoadedBinders = LoadedBinders
-  { lbBinders :: [(CoreSyn.CoreBndr, CoreSyn.CoreExpr)]
+  { lbBinders :: !(Map CoreSyn.CoreBndr CoreSyn.CoreExpr)
   -- ^ Binder + expression it's binding
-  , lbClassOps :: [(CoreSyn.CoreBndr, Int)]
+  , lbClassOps :: !(Map CoreSyn.CoreBndr Int)
   -- ^ Type class dict projection functions
-  , lbUnlocatable :: [CoreSyn.CoreBndr]
+  , lbUnlocatable :: !(Set CoreSyn.CoreBndr)
   -- ^ Binders with missing unfoldings
-  , lbPrims :: [Either UnresolvedPrimitive FilePath]
+  , lbPrims :: !(Seq (Either UnresolvedPrimitive FilePath))
   -- ^ Primitives; either an primitive data structure or a path to a directory
   -- containing json files
-  , lbReprs :: [DataRepr']
+  , lbReprs :: !(Seq DataRepr')
   -- ^ Custom data representations
+  , lbCache :: !DeclCache
+  -- ^ Loaded module cache
   }
 
-mergeLoadedBinders :: [LoadedBinders] -> LoadedBinders
-mergeLoadedBinders lbs =
-  LoadedBinders {
-    lbBinders=concat (map lbBinders lbs)
-  , lbClassOps=concat (map lbClassOps lbs)
-  , lbUnlocatable=concat (map lbUnlocatable lbs)
-  , lbPrims=concat (map lbPrims lbs)
-  , lbReprs=concat (map lbReprs lbs)
-  }
+type LoadedBinderT m a = StateT LoadedBinders m a
+
+-- | Stores modules with easy binder lookup
+type DeclCache = Map GHC.Module (Maybe (Map GHC.Name IfaceSyn.IfaceDecl))
+
+-- | Insert a list of keys and values into a 'Map'
+mapInsertAll :: Ord k => Map k a -> [(k, a)] -> Map k a
+mapInsertAll = foldl' (\m (k, v) -> Map.insert k v m)
```

Comment:
You only use this in one place, it may as well be inlined

Reply:
Alright, I've moved it to be a local definition. I've kept it as a separate thing, because I think `mapInsertAll m xs` is a bit more parsable than `foldl' (\m (k, v) -> Map.insert k v m) theMap xs`